### PR TITLE
[DPE-1779] COS

### PR DESCRIPTION
**Pull request by dragomirp**
## Issue
Add COS integration

## Solution
* Add an exporter service
* Add a graphana dashboard definition
* Add COS relations

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -66,9 +72,22 @@ def __init__(self, *args):
...
         ]
+        self._metrics_service = "metrics_server"
+        self.grafana_dashboards = GrafanaDashboardProvider(self)
+        self.metrics_endpoint = MetricsEndpointProvider(
+            self,
+            jobs=[{"static_configs": [{"targets": [f"*:{METRICS_PORT}"]}]}],
+        )
+        self.loki_push = LogProxyConsumer(
+            self,
+            log_files=[f'{service["log_dir"]}/pgbouncer.log' for service in self._services],
```
Comment (dragomirp) [inline src/charm.py:87]: In the future we should most probably listen to pebble instead.

File: src/charm.py [hunk 2]
```diff
@@ -294,6 +324,40 @@ def reload_pgbouncer(self) -> None:
 
         self.check_pgb_running()
 
+    def _generate_monitoring_service(self, enabled: bool = True) -> Dict[str, str]:
+        if enabled:
+            stats_password = self.peers.get_secret("app", MONITORING_PASSWORD_KEY)
+            command = (
+                f'pgbouncer_exporter --web.listen-address=:{METRICS_PORT} --pgBouncer.connectionString="'
+                f'postgres://{self.backend.stats_user}:{stats_password}@localhost:{self.config["listen_port"]}/pgbouncer?sslmode=disable"'
```
Comment (dragomirp) [inline src/charm.py:332]: Should we be using TLS here if enabled? It is a local service, so it shouldn't matter.

File: src/grafana_dashboards/pgbouncer-metrics.json
Comment (dragomirp) [inline src/grafana_dashboards/pgbouncer-metrics.json:1]: Copied over from the VM charm

File: src/relations/peers.py [hunk 3]
```diff
@@ -222,7 +221,8 @@ def _on_changed(self, event: RelationChangedEvent):
             try:
                 # raises an error if this is fired before on_pebble_ready.
                 self.charm.reload_pgbouncer()
-            except ConnectionError:
+                self.charm.toggle_monitoring_layer(self.charm.backend.ready)
+            except (ConnectionError, ChangeError):
```
Comment (dragomirp) [inline src/relations/peers.py:225]: If the `listen_port` is changed, pebble cannot know and the monitoring service may fail before a PGB service is available on the new port.

**Review by taurus-forever — APPROVED** [review]
Tested as PPA and it does the job and looks great! Thank you!
BTW, there are some issues with pgbouncer removal, but not related to COS, let's report and address them separately. Good job done!

File: src/charm.py (hunk 2 above)
Comment (marceloneppel) [reply src/charm.py:332]: From what Mykola told me sometime ago, all the connections to the database should use TLS when available, but I think all the tools in the PG charm are using IP addresses on VM and hostnames on K8S, so I'm not sure whether it's needed for localhost. It looks that it's not worth locally.

**Review by marceloneppel — APPROVED** [review]

File: src/grafana_dashboards/pgbouncer-metrics.json [hunk 4]
```diff
@@ -0,0 +1,1244 @@
...
+      "24h",
+      "2d",
+      "7d",
+      "30d"
+    ]
+  },
+  "timezone": "browser",
+  "title": "PgBouncer",
+  "uid": "aek-7cbd6b6f87",
+  "version": 2
```
[the remark this answers is not in the document]
Comment (dragomirp) [reply src/grafana_dashboards/pgbouncer-metrics.json]: Updated title, uid and tags. So dashes are available in parallel. Grafana can still display VM pgb in the K8S dashboards and vice versa.

**Pull request merged**